### PR TITLE
[Nodejs] initial integration.

### PR DESCRIPTION
**Pull request by DavidKorczynski**
Fuzzing for nodejs core. Notice that the build time here can take quite some time ~10/15 minutes, since we need to build v8 first.

The `project.yaml` file has been here since December 2016 so I am not sure if the same information is valid anymore. The original `project.yaml` was added by @oliverchang so perhaps you might have an idea here? I have not reached out to the security people at Nodejs as I assume if the contact information is accurate, if not then please do tell me.

### CODE REVIEW

**Comment by inferno-chromium** [comment]
@mikeal, can you please review this. Also, can you suggests some important places to fuzz, this one feels like a simple url parsing example.

**Comment by DavidKorczynski** [comment]
Note that the failures of the builds are due to timeouts in Travis.

**Comment by catenacyber** [comment]
Any plans on a target taking javascript as input ?

**Comment by inferno-chromium** [comment]
> Any plans on a target taking javascript as input ?

Yes i think those would be good targets. Still we need maintainer approval first, asking @mikeal @bnoordhuis @ry

**Comment by DavidKorczynski** [comment]
> Any plans on a target taking javascript as input ?

Yeah - my approach on Nodejs will be to cover it extensively.

@inferno-chromium I have also reached out to the security team of nodejs and am waiting to hear back. Should have something soon.

**Comment by DavidKorczynski** [comment]
@aaddaleax or @jasnell  could you maybe assist here? The set up in this PR is used to find the bug in https://github.com/nodejs/node/pull/33640 and it would be nice to get some responds on whether the NodeJS maintainers would be happy to have fuzzing of NodeJS integrated via OSS-Fuzz?

**Comment by DavidKorczynski** [comment]
> LGTM insofar I'm able to judge that. Thanks, David.

Thanks a lot - this one took quite a bit more effort!

> I'm open to extending node's build to make it easier to integrate with this project. That build.sh script looks like it could be a lot of work to maintain.

I would be happy to assist here and can give references on the difficulties that occurred when integrating it into the OSS-Fuzz environment. Should we do it over an issue on Nodejs?

**Comment by bnoordhuis** [comment]
Yes, good idea.

**Comment by inferno-chromium** [comment]
@oliverchang - can you please help to review Nodejs, we are very excited to accept this project.

File: projects/nodejs/fuzz_url.cc [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include "env-inl.h"
+#include "node_crypto.h"
+#include "node_crypto_common.h"
+#include "node.h"
+#include "node_internals.h"
+#include "node_url.h"
+#include "string_bytes.h"
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+    node::url::URL url2((char*)data, size);
```
Comment (oliverchang) [inline projects/nodejs/fuzz_url.cc:26]: this target is pretty basic. do you have plans for targets that test more core nodejs functionality?

**Comment by jasnell** [comment]
I likely won't be able to take a detailed look at this for at least a few more days but in general: yes, I'd love to see some regular fuzz testing for Node.js

**Comment by inferno-chromium** [comment]
> LGTM insofar I'm able to judge that. Thanks, David.
> 
> I'm open to extending node's build to make it easier to integrate with this project. That build.sh script looks like it could be a lot of work to maintain.

Yes build.sh is extremely complicated right now, can you please make it a proper build config upstream (optionally if it can be tested in CI, that is super nice). Then we can merge this.

File: projects/nodejs/build.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# limitations under the License.
+#
+################################################################################
+cd node
+
+export LDFLAGS="$CXXFLAGS"
+export LD="$CXX"
+
+./configure --without-intl --without-node-code-cache --without-dtrace --without-snapshot --without-ssl
+make -j4
```
Comment (intrigus-lgtm) [inline projects/nodejs/build.sh:23]: Maybe change this to `-j$(nproc)` to get faster builds, when manually building this?

Comment (DavidKorczynski) [reply projects/nodejs/build.sh]: Agreed

**Pull request merged**